connector/connect/common/src/main/protobuf/spark/connect/base.proto

```diff
@@ -539,6 +558,138 @@ message AddArtifactsResponse {
+
+    // (Required) The operation request id to be interrupted.
+    string operation_id = 4;
+}
+
+// TODO - depracated by ReleaseOperationRequest becomes reudndant
+message InterruptResponse {
+  string session_id = 1;
+  string operation_id = 2;
+}
+
+message ReleaseOperationRequest {
+  // (Required)
+  //
+  // The session_id specifies a spark session for a user id (which is specified
+  // by user_context.user_id). The session_id is set by the client to be able to
+  // collate streaming responses from different queries within the dedicated session.
+  string session_id = 1;
+
+  // (Required) User context
+  UserContext user_context = 2;
+
+  // Provides optional information about the client sending the request. This field
+  // can be used for language or version specific information and is only intended for
+  // logging purposes and will not be interpreted by the server.
+  optional string client_type = 3;
+
+  // (Required) The operation request id to be interrupted.
+  string operation_id = 4;
+
+  // When true, all data about the operation can be fully released from the server,
+  // and any executing operation can be interrupted.
+  bool remove = 5;
+
+  // (Optional) If remove is false, this can be used to instruct the server to interrupt the
+  // operation execution.
+  optional bool interrupt = 6;
+
+  // (Optional) If remove is false, this can be used to instruct the server that query
+  // below this offset (exclusive) will no longer be requested by the client.
+  optional int64 free_query_results_below_offset = 7;
+}
+
+message ReleaseOperationResponse {
+  string session_id = 1;
+  string operation_id = 2;
+
+  // When true, the operation has been fully released from the server.
+  bool removed = 3;
+
+  // When true, the operation successfully finished executing,
+  // so any effects of the operation have been committed.
+  bool executed = 4;
+}
+
+message ReleaseSessionRequest {
+  // (Required)
+  //
+  // The session_id specifies a spark session for a user id (which is specified
+  // by user_context.user_id). The session_id is set by the client to be able to
+  // collate streaming responses from different queries within the dedicated session.
+  string session_id = 1;
+
+  // (Required) User context
+  UserContext user_context = 2;
+
+  // Provides optional information about the client sending the request. This field
+  // can be used for language or version specific information and is only intended for
+  // logging purposes and will not be interpreted by the server.
+  optional string client_type = 3;
+
+  // When true, release all server state of this session (interrupting all executions).
+  bool remove = 4;
+
+  // This leaves place for extensions of this request, where with remove=false, clients
+  // can request
+  // E.g. release all operations that finished executing.
+}
+
+message ReleaseSessionResponse {
+  string session_id = 1;
+  bool removed = 2;
+}
+
+message PingRequest {
```

Comment:
Why have an explicit ping instead of https://github.com/grpc/grpc/blob/master/doc/keepalive.md?

Reply:
In DBSQL, I kind of regret the inactivity timeout. It's caused a bunch of headaches when there is network connectivity issues. We're slowing moving to a model where we expose EXECUTION_TIMEOUT and RESULT_RETENTION as two static numbers for how long the query can run for and how long results are fetchable for. Thoughts?

Reply:
So the meaning of something like this instead of using GRPC channel keepalive settings would be exactly to decouple this retention of the session from timeout of the connection. E.g. a server could send this PingRequest with a hearbeat_timeout = 24 hours, to be able to disconnect and keep the session state until the next day. Or set it to 5 minutes and this way tell the server that soon after the client stops pinging, the server may release state.
Maybe naming it after "retention" rather than ping and hearbeat would be clearer.

With GRPC keepalives, I am also not sure if/how to get to this information on the server side, so that server can e.g. throw away the session based on lack of pings. https://github.com/grpc/grpc-java/issues/779 suggests that there is no way for a netty server to get information about the open grpc channels.